### PR TITLE
869 add distributed training support for AUC metric

### LINKED ISSUE #869: Add distributed support for ROC_AUC metric
**Is your feature request related to a problem? Please describe.**
As ignite doesn't support distributed training for `epoch-gather` metrics, we need to enhance `ROCAUC` referring to:
https://pytorch.org/docs/stable/distributed.html#torch.distributed.gather

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #869 .

### Description
This PR added distributed training support to the AUC ignite metric / handler.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/handlers/roc_auc.py [hunk 1]
```diff
@@ -96,6 +98,13 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
     def compute(self):
         _prediction_tensor = torch.cat(self._predictions, dim=0)
         _target_tensor = torch.cat(self._targets, dim=0)
+
+        if dist.is_available() and dist.is_initialized() and not self._is_reduced:
+            dist.barrier()
```
Comment (vfdev-5) [inline monai/handlers/roc_auc.py:103]: Maybe, no need to use barrier here as all_gather will sync

File: monai/handlers/roc_auc.py [hunk 2]
```diff
@@ -96,6 +98,13 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
...
+            _prediction_tensor = dist.gather(_prediction_tensor)
```
Comment (vfdev-5) [inline monai/handlers/roc_auc.py:104]: `dist.gather` -> `dist.all_gather`

File: monai/handlers/roc_auc.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:103]: updated in the latest commit.
Thanks.

File: monai/handlers/roc_auc.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:104]: Thanks for correcting me, updated in the latest commit.

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

I updated the PR according to your comments and sample code, could you please help review it again?
Thanks.

**Comment by vfdev-5** [comment]
@Nic-Ma it would be better to provide some tests, I think. How do you guys test distributed features ?

File: monai/handlers/roc_auc.py [hunk 3]
```diff
@@ -96,6 +98,17 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
     def compute(self):
         _prediction_tensor = torch.cat(self._predictions, dim=0)
         _target_tensor = torch.cat(self._targets, dim=0)
+
+        if dist.is_available() and dist.is_initialized() and not self._is_reduced:
+            # create placeholder to collect the data from all processes:
+            output = [torch.zeros_like(_prediction_tensor) for _ in range(dist.get_world_size())]
+            dist.all_gather(output, _prediction_tensor)
```
Comment (vfdev-5) [inline monai/handlers/roc_auc.py:105]: Here it may fail if `_prediction_tensor` is not on the correct device according to dist backend. For example, nccl <-> "cuda:local_rank" or gloo <-> "cpu"

**Comment by Nic-Ma** [comment]
> @Nic-Ma it would be good to provide some tests, I think. How do you guys test distributed features ?

You are right. Actually, this is the first distributed training related feature(others are examples), we haven't built tests.
I will try to test locally at least. Does the code look reasonable now?

Thanks.

File: monai/handlers/roc_auc.py [hunk 4]
```diff
@@ -96,6 +98,17 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
...
+            _prediction_tensor = torch.cat(output, dim=0)
+            output = [torch.zeros_like(_target_tensor) for _ in range(dist.get_world_size())]
+            dist.all_gather(output, _target_tensor)
+            _target_tensor = torch.cat(output, dim=0)
+            self._is_reduced = True
+
         return compute_roc_auc(
```
Comment (vfdev-5) [inline monai/handlers/roc_auc.py:112]: Here, either `compute_roc_auc` should know how to deal with tensors on CUDA (if nccl dist backend), either we need to put them to the CPU before. And here, all processess will execute on the CPU the same computation which maybe a bit redundant. That's why I talked about broadcasting in ignite's issue.

File: monai/handlers/roc_auc.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:105]: I see, thanks for your reminder.

**Comment by vfdev-5** [comment]
> You are right. Actually, this is the first distributed training related feature(others are examples), we haven't built tests.
> I will try to test locally at least.

I see. Part of features you may easily test even without having multiple GPUs using "gloo" backend and spawning multiple processes with `torch.distributed.launch`, for example. This can be even run on Github Actions CI etc.

File: monai/handlers/roc_auc.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:112]: `compute_roc_auc` can handle CUDA tensor here.
As you said in ignite thread, maybe `all_gather` is compatible for all kinds of backends, I think we can use this logic for the first version.
Thanks.

Comment (vfdev-5) [reply monai/handlers/roc_auc.py:112]: > As you said in ignite thread, maybe all_gather is compatible for all kinds of backends, I think we can use this logic for the first version.

Just to detail that: `all_gather`/`gather` works for all torch native backends. In ignite thread (just added it), I was talking about ignite supported backend (torch native, xla, horovod). So, Horovod and XLA do not have `gather` collective operation, that's why I proposed to keep `all_gather`.

Comment (Nic-Ma) [reply monai/handlers/roc_auc.py:112]: Thanks for your explanation!

**Comment by Nic-Ma** [comment]
Hi @wyli and @ericspod ,

I executed the test program locally:
```
python -m torch.distributed.launch --nproc_per_node=2 --nnodes=1
       --node_rank=0 --master_addr="10.23.137.29"
       --master_port=1234 test_handler_rocauc_dist.py
```
And it passed as expected.
Could you please help review this PR?
I think we can easily add the test program to CI later when @IsaacYangSLA set up CI test env for distributed training.

Thanks.
